airbyte-integrations/connectors/source-datadog/source_datadog/streams.py

```diff
@@ -0,0 +1,288 @@
+            "page": {"limit": self.limit},
+        }
+        if cursor:
+            payload["page"]["cursor"] = cursor
+
+        return payload
+
+
+class AuditLogs(IncrementalSearchableStream):
+    """
+    https://docs.datadoghq.com/api/latest/audit/#search-audit-logs-events
+    """
+
+    def path(self, **kwargs) -> str:
+        return "audit/events/search"
+
+
+class Logs(IncrementalSearchableStream):
+    """
+    https://docs.datadoghq.com/api/latest/logs/#search-logs
+    """
+
+    def path(self, **kwargs) -> str:
+        return "logs/events/search"
+
+
+class BasedListStream(V2ApiDatadogStream, ABC):
+    parse_response_root: Optional[str] = "data"
+
+    @property
+    def http_method(self) -> str:
+        return "GET"
+
+
+class Metrics(BasedListStream):
+    """
+    https://docs.datadoghq.com/api/latest/metrics/#get-a-list-of-metrics
+    """
+
+    def path(self, **kwargs) -> str:
+        return "metrics?window[seconds]=1209600"  # max value allowed (2 weeks)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+
+class PaginatedBasedListStream(BasedListStream, ABC):
+    primary_key: Optional[str] = "id"
+
+    def path(
+        self,
+        *,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        offset = None
+        if next_page_token:
+            offset = next_page_token.get("offset")
+        return self.get_url_path(offset)
+
+    @abstractmethod
+    def get_url_path(self, offset: Optional[str]) -> str:
+        """
+        Returns the relative URL with the corresponding offset
+        """
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        response_json = response.json()
+        next_offset = response_json.get("meta", {}).get("pagination", {}).get("next_offset", -1)
+        current_offset = response_json.get("meta", {}).get("pagination", {}).get("offset", -1)
+        next_page_token = None
+        if next_offset != current_offset:
+            next_page_token = {"offset": next_offset}
+        return next_page_token
+
+
+class Incidents(PaginatedBasedListStream):
+    """
+    https://docs.datadoghq.com/api/latest/incidents/#get-a-list-of-incidents
+    """
+
+    def get_url_path(self, offset: Optional[str]) -> str:
+        params = f"&page[offset]={offset}" if offset else ""
+        return f"incidents?page[size]={self.limit}{params}"
+
+
+class IncidentTeams(PaginatedBasedListStream):
+    """
+    https://docs.datadoghq.com/api/latest/incident-teams/#get-a-list-of-all-incident-teams
+    """
+
+    def get_url_path(self, offset: Optional[str]) -> str:
+        params = f"&page[offset]={offset}" if offset else ""
+        return f"teams?page[size]={self.limit}{params}"
```

Comment:
Can you add a unit test for this?

Reply:
btw why this functino is not moved to PaginatedBasedListStream?

Reply:
Because the same logic doesn't apply for `Users`